{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- the purpose of this Py script is to count separately the number of human vs Neanderthal loci found in the sequencing data. \n",
    "- also to conclude if the sampled individual is related closer to modern humans or to Neanderthals.\n",
    "\n",
    "*Graphical abstract for the code* |  *The pseudocode for the script*|\n",
    ":-----:     |:--------------:|\n",
    "<img src=\"graphical_abstr_count1.jpg\" alt=\"graphical abstract\" title=\"Graphical abstract for the code\" width=\"400\" /> | <img src=\"script_pseudocode.jpg\" alt=\"pseudocode\" title=\"Graphical abstract for the code\" width=\"700\" /> |"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"script_pseudocode1.jpg\" width=\"400\" />   \n",
    "*More pseudocode example* "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gather the necessary files\n",
    "* on Uppmax, the original files are in:  \n",
    "/proj/snic2019-8-150/nobackup/private/Mirela/neanderthal_project/data/raw_internal  \n",
    "/home/miba8458/LnktoWorkingFolder/UppmaxDataLnk/data/raw_external/  \n",
    "\n",
    "* to work on them, I have copied the files to this location  \n",
    "/home/miba8458/LnktoWorkingFolder    \n",
    "/home/miba8458/private/workingFolder  \n",
    "\n",
    "* the orinal files to be used as control (known if the aDNA they contain was either human or Neanderthal)  \n",
    "/proj/snic2019-8-150/nobackup/private/Mirela/neanderthal_project/data/raw_external  \n",
    "\n",
    "[ ] the .bam sequencing files, filtered and sorted  \n",
    "[ ] the .ibam files, containing the indexes of the .bam files  \n",
    "[ ] the list with the genomic locations known to differ between modern humans and Neanderthals   \n",
    "  (#Ref paper: Phylogenetic analysis in 'Nuclear DNA sequences from the Middle Pleistocene Sima de los Huesos hominins' - Meyer, Nature, 2016    \n",
    "   #DL link: https://bioinf.eva.mpg.de/sima/)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## explore the files \n",
    "- to see what is needs to be extracted  \n",
    "### A) the reference list:   \n",
    "    - col 1 - chr nr (chromosome on hg19/GRCh37) \n",
    "    - col 2 - genomic location (position on hg19/GRCh37) - unique position?\n",
    "    - col 3 - ancestral state present in the chimpanzee, bonobo, gorilla, orangutan and rhesus macaque reference sequences  \n",
    "    - col 4 - derived state  \n",
    "    - col 5 - of which species is the substitution representative of (indicates the presence of the derived state Altai Neandertal, Denisovan or Mbuti. \"all\" means that all three were derived)   \n",
    "        - Neandertal  \n",
    "        - Denisova  \n",
    "        - Human  (<- Mbuti genome)\n",
    "        - Human-Denisova   \n",
    "        - Neandertal-Human   \n",
    "        - Neandertal-Denisova  \n",
    "        - all  \n",
    "  \n",
    "(to obtain the values of col 5: `less -c sima_informative_sites.tab | awk '{a[$5]++} END {for (n in a) print n, a[n]}'`)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#bash code  \n",
    "#do this in Py when having time   \n",
    "`less sima_informative_sites.tab | head -5` - to read the file and display the first 5 rows  \n",
    "10\t94488\tG\tA\tall\t \n",
    "10\t95365\tT\tC\tall\t \n",
    "10\t95830\tT\tC\tall\t   \n",
    "10\t97555\tG\tA\tall\t \n",
    "10\t97647\tG\tT\tall\t \n",
    "\n",
    "`sed -n '55,60p' <sima_informative_sites.tab` - to display the rows 55 to 60 of the file   \n",
    "10\t147482\tT\tC\tNeandertal \t\n",
    "10\t147544\tG\tA\tall\t \n",
    "10\t147812\tA\tG\tall\t \n",
    "10\t148000\tA\tG\tall\t \n",
    "10\t148199\tC\tT\tall\t \n",
    "10\t148542\tG\tA\tDenisova\t\t "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### B) the file containing the genome reference \n",
    "* we are using the hs37d5.fa reference genome\n",
    "* When I printed out the first few lines, I got lots of 'NNNNNNNNNNNNN'. Actually lines 2-168 are only like that.\n",
    "* looked further down, and I printed lines 1 and 170:  \n",
    "`zless hs37d5.fa.gz|sed -n '1p;170p`\n",
    ">1 dna:chromosome chromosome:GRCh37:1:1:249250621:1\n",
    "ACCCTAACCCTAACCCTAACCCTAACCCAACCCTAACCCTAACCCTAACCCTAACCCTAA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### C) the files containing the sequences\n",
    "- .bam files with 11 mandatory columns and 1 optional\n",
    "- in the .bam file, the columns that I am interested in are:\n",
    "    - col 3: reference sequence  \n",
    "    - col 4: position number (the starting position of the read)  \n",
    "    - col 6: alphanumeric; if only 'M', 'I', 'S', '=' and/ or 'X', then the length of the sequence is the sum of the numbers in this string   \n",
    "  \n",
    "(to look at the first 5 rows of the *alignment* and ignore the header of the .bam:    \n",
    "`samtools view jar001-b1e1l1p1_CATGCTC_L004_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam | head -5`)  \n",
    "    \n",
    "(to print the min and max values of col 4 - in the 2 .bam files I have:   \n",
    "`samtools view jar001-b1e1l1p1_CATGCTC_L004_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam | awk 'BEGIN{a=1000}{if ($4<0+a) a=$4} END{print a}'`  \n",
    "result - 5  | 178  \n",
    "`samtools view jar001-b1e1l1p1_CATGCTC_L004_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam | awk 'BEGIN{a=   0}{if ($4>0+a) a=$4} END{print a}'`  \n",
    "result - 249 240 549 | 249 239 780)  \n",
    "   \n",
    "(to look at a few values around the max value of col 4, I change the value and use it as a int instead of a string:)  \n",
    "`samtools view jar002-b1e1l1p1_ACGCAAC_L005_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam | awk '{if ($4 > 249200000) print $0;}'`  \n",
    "  \n",
    "(to look at the total number of reads:)\n",
    "`samtools idxstats jar001-b1e1l1p1_CATGCTC_L004_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam| awk '{s+=$3+$4} END {print s}'`\n",
    "   \n",
    "(first row of the .bam file)  \n",
    "M_ST-E00201:191:HH7YYALXX:4:1103:31548:29806\t16\t1\t10022\t0\t79M\t*\t0\t0\tCCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAAC\t]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]Z]]]]]]]]]]\tXT:A:R\tNM:i:0\tX0:i:370\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:79\tXP:i:1  \n",
    "\n",
    "\n",
    "* get data out of the .bam files:\n",
    "`samtools mpileup -f hs37d5.fa.gz jar001-b1e1l1p1_CATGCTC_L004_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam | head -5`  \n",
    "\\[mpileup\\] 1 samples in 1 input files  \n",
    "1\t10022\tC\t1\t^!,\tA  \n",
    "1\t10023\tC\t1\t,\tB  \n",
    "1\t10024\tC\t1\t,\tE  \n",
    "1\t10025\tT\t1\t,\t]  \n",
    "1\t10026\tA\t1\t,\t]      \n",
    "  \n",
    "col1 - Sequence name (chr)  \n",
    "col2 - 1-based coordinate  \n",
    "col3 - Reference base  \n",
    "col4 - Number of reads covering this position  \n",
    "col5 - Read bases  \n",
    "col6 - Base qualities  \n",
    "\n",
    "\n",
    "### ~~Making it a .vcf file with `mpileup`~~  \n",
    "`samtools mpileup -f hs37d5.fa.gz -g jar001-b1e1l1p1_CATGCTC_L004_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam|bcftools call -m| sed -n '114,117p'`  \n",
    "Note: none of --samples-file, --ploidy or --ploidy-file given, assuming all sites are diploid  \n",
    "\\[warning\\] samtools mpileup option `g` is functional, but deprecated. Please switch to using bcftools mpileup in future.  \n",
    "\\[mpileup\\] 1 samples in 1 input files  \n",
    "\n",
    "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT  \t\n",
    "1\t10022\t.\tC\t.\t33.5884\t.\tDP=1;MQ0F=1;AN=2;DP4=0,1,0,0;MQ=0\tGT\t0/0  \n",
    "1\t10023\t.\tC\t.\t33.5884\t.\tDP=1;MQ0F=1;AN=2;DP4=0,1,0,0;MQ=0\tGT\t0/0  \n",
    "1\t10024\t.\tC\t.\t33.5884\t.\tDP=1;MQ0F=1;AN=2;DP4=0,1,0,0;MQ=0\tGT\t0/0  \n",
    "\n",
    "### ~~Variant calling with `bcftools` and getting a .vcf~~   \n",
    "* write the output to 'test.vcf.gz'  \n",
    "`bcftools mpileup -Ou -f hs37d5.fa.gz jar001-b1e1l1p1_CATGCTC_L004_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam|bcftools call -vmO z -o test.vcf.gz`  \n",
    "Note: none of --samples-file, --ploidy or --ploidy-file given, assuming all sites are diploid  \n",
    "\\[mpileup\\] 1 samples in 1 input files  \n",
    "* read lines 114-117  \n",
    "`less test.vcf.gz| sed -n '114,117p'`  \n",
    "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT \t \n",
    "1\t649113\t.\tC\tT\t8.13869\t.\tDP=1;SGB=-0.379885;MQ0F=0;AC=2;AN=2;DP4=0,0,0,1;MQ=37\tGT:PL\t1/1:37,3,0  \n",
    "1\t861630\t.\tG\tA\t8.13869\t.\tDP=1;SGB=-0.379885;MQ0F=0;AC=2;AN=2;DP4=0,0,0,1;MQ=37\tGT:PL\t1/1:37,3,0  \n",
    "1\t918384\t.\tG\tT\t8.13869\t.\tDP=1;SGB=-0.379885;MQ0F=0;AC=2;AN=2;DP4=0,0,0,1;MQ=37\tGT:PL\t1/1:37,3,0  \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## C) What data I need  \n",
    "- from the list with genomic locations:  \n",
    "    - list 1: (Neanderthal & Neanderthal-Denisova) locations\n",
    "    - list 2: (Human & Human-Denisova) locations\n",
    "    \n",
    "*** ADD SCHEMATIC HERE ***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### D) Getting everything together  \n",
    "#### 1) copy the reference genome from my PC to Uppmax:  \n",
    "* Local terminal (**don't** log in Uppmax) > `scp address/local_file user@rackham.uppmax.uu.se:/home/usernameRackham/folder`   \n",
    "#### 2) log into Uppmax & check file location and integrity  \n",
    "==LOCAL WORK NOT JOB==\n",
    "#### 3) load necessary modules into Uppmax  \n",
    "* to see available modules: `module avail`  \n",
    "* to load modules, such as `samtools`, load `bioinfo-tools` first:  \n",
    "\t`module load bioinfo-tools samtools`  \n",
    "#### 4) Get the necessary info from .bam files:  \n",
    "* the .bam files contain strings of DNA with the start location of each string. What I need is the precise location of each nucleotide. That information is obtained by using `samtools mpileup` which extracts this information, besides other details.  \n",
    "`samtools mpileup -f hs37d5.fa.gz jar001-b1e1l1p1_CATGCTC_L004_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam > jar001.mpileup`  \n",
    "and  \n",
    "`samtools mpileup -f hs37d5.fa.gz jar002-b1e1l1p1_ACGCAAC_L005_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam > jar002.mpileup`  \n",
    "\n",
    "==UPPMAX JOB==  \n",
    "#### 3) do the script to submit a job, then copy it on Uppmax  \n",
    "++++++++++++++++++++++++++++++++++  \n",
    "`#!/bin/bash -l  \n",
    "#SBATCH -A snic2019-8-150  \n",
    "#SBATCH -t 22:00:00  \n",
    "#SBATCH -p core  \n",
    "  \n",
    "\\# load modules  \n",
    "module load bioinfo-tools samtools  \n",
    "  \n",
    "\\# run mpileup to get the necessary positional information  \n",
    "samtools mpileup -f /home/miba8458/private/workingFolder/raw_internal/hs37d5.fa.gz /home/miba8458/private/workingFolder/raw_internal/jar001-b1e1l1p1_CATGCTC_L004_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam > /home/miba8458/private/workingFolder/raw_internal/jar001.mpileup`  \n",
    "\n",
    "++++++++++++++++++++++++++++++++++  \n",
    "* name the file and give it the extension '.sh'  \n",
    "* (somewhere it was written to make it executable: chmod u+x file.sh, but I did it without it)  \n",
    "* stay on your local drive and do point 1) again  \n",
    "* go on Uppmax  \n",
    "* submit job: `sbatch jar001mpileup.sh `  \n",
    "* you get an ID for the job, in this case 11608594 and for second file: 11609004\n",
    "  \n",
    "#### 4) check the job  \n",
    "`squeue -u miba8458`  \n",
    "             JOBID PARTITION     NAME     USER ST       TIME  NODES NODELIST(REASON)  \n",
    "          11596631      core jar001mp miba8458 PD       0:00      1 (Nodes required for job are DOWN, DRAINED or reserved for jobs in higher priority partitions)  \n",
    "\n",
    "`jobinfo -u miba8458`  \n",
    "Waiting jobs:  \n",
    "   JOBID    POS PARTITION                      NAME     USER        ACCOUNT ST          START_TIME   TIME_LEFT PRIORITY CPUS NODELIST(REASON)     FEATURES DEPENDENCY  \n",
    "11596631    823      core          jar001mpileup.sh miba8458 snic2019-8-150 PD                 N/A    22:00:00   100010    1 (Nodes required        (null)   \n",
    "\n",
    "\n",
    "#### 5) jobs are done  \n",
    "* check the slurm log files to see how the job went, if there was any problem\n",
    "`less`\n",
    "* check the output files  \n",
    "`less`  \n",
    "* see their line nr, etc\n",
    "``  \n",
    "* copy files from Uppmax to your PC  \n",
    "`scp user@rackham.uppmax.uu.se:/home/username/some-file local_file_or_directory`  \n",
    "(To place the file in the directory you are currently standing in, use a dot ('.') as the local directory:  \n",
    "`scp user@milou.uppmax.uu.se:/home/username/some-file .`)    \n",
    "\n",
    "#### ~~3) try to use the node of the own HDD~~\n",
    "https://www.uppmax.uu.se/support/user-guides/how-to-use-the-nodes-own-hard-drive-for-analysis/  \n",
    "* rewrote the code:  \n",
    "\n",
    "+++++++++++++++++++++++++++++++  \n",
    "`#!/bin/bash -l  \n",
    "#SBATCH -A snic2019-8-150  \n",
    "#SBATCH -t 22:00:00  \n",
    "#SBATCH -p core  \n",
    "  \n",
    "\\# load modules  \n",
    "module load bioinfo-tools samtools  \n",
    "  \n",
    "\\# copy the files used in the analysis to \\$SNIC\\_TMP  \n",
    "cp /home/miba8458/private/workingFolder/raw_internal/hs37d5.fa.gz /home/miba8458/private/workingFolder/raw_internal/jar001-b1e1l1p1_CATGCTC_L004_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam $SNIC_TMP  \n",
    "  \n",
    "cp /home/miba8458/private/workingFolder/raw_internal/hs37d5.fa.gz /home/miba8458/private/workingFolder/raw_internal/jar002-b1e1l1p1_ACGCAAC_L005_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam $SNIC_TMP  \n",
    "  \n",
    "\\# go to the $SNIC_TMP folder to make sure any temporary files are created there as well  \n",
    "cd $SNIC_TMP  \n",
    "  \n",
    "\\# run mpileup to get the necessary positional information  \n",
    "samtools mpileup -f $SNIC_TMP/hs37d5.fa.gz $SNIC_TMP/jar001-b1e1l1p1_CATGCTC_L004_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam > $SNIC_TMP/jar001TMP.mpileup  \n",
    "  \n",
    "samtools mpileup -f $SNIC_TMP/hs37d5.fa.gz $SNIC_TMP/jar001-b1e1l1p1_CATGCTC_L004_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam -> $SNIC_TMP/jar002TMP.mpileup  \n",
    "  \n",
    "\\# copy the results back to the network file system  \n",
    "cp $SNIC_TMP/jar001TMP.mpileup /home/miba8458/private/workingFolder/raw_internal/  \n",
    "cp $SNIC_TMP/jar002TMP.mpileup /home/miba8458/private/workingFolder/raw_internal/  \n",
    "  \n",
    "+++++++++++++++++++++++++++++++  \n",
    "11596876\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### G) Check the 3 needed files:  \n",
    "* the list with genomic locations:  \n",
    "    * nr of lines:  \n",
    "    `less -c sima_informative_sites.tab |wc -l`  \n",
    "    or  \n",
    "    `less -c sima_informative_sites.tab |awk 'END { print NR }'`\n",
    "    7 049 547\n",
    "    * first 5 lines:  \n",
    "    10\t94488\tG\tA\tall  \t\n",
    "    10\t95365\tT\tC\tall  \t\n",
    "    10\t95830\tT\tC\tall\t \n",
    "    10\t97555\tG\tA\tall\t \n",
    "    10\t97647\tG\tT\tall\t \n",
    "    * last 5 lines:  \n",
    "    9\t140991238\tG\tA\tall\t \n",
    "    9\t140992405\tA\tG\tall\t \n",
    "    9\t140992739\tC\tT\tDenisova  \n",
    "    9\t140992743\tT\tC\tall\t \n",
    "    9\t140993341\tC\tT\tall\t \n",
    "* the sequencing lists:  \n",
    "    * load the samtools module if you are on Uppmax:  \n",
    "    `module load bioinfo-tools samtools`  \n",
    "    * nr of lines (without header!):  \n",
    "    `samtools view jar001-b1e1l1p1_CATGCTC_L004_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam | wc -l`  \n",
    "    or  \n",
    "    `samtools view jar001-b1e1l1p1_CATGCTC_L004_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam |awk 'END { print NR }'`  \n",
    "    469 379 - for 1st    \n",
    "    and  \n",
    "    65 691 - for 2nd  \n",
    "    * first 5 lines for jar001:  \n",
    "    `samtools view jar001-b1e1l1p1_CATGCTC_L004_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam | head -n 5`  \n",
    "    M_ST-E00201:191:HH7YYALXX:4:1103:31548:29806\t16\t1\t10022\t0\t79M\t*\t0\t0\tCCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAACCCTAAC\t]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]Z]]]]]]]]]]\tXT:A:R\tNM:i:0\tX0:i:370\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:79\tXP:i:1  \n",
    "    \n",
    "    M_ST-E00201:191:HH7YYALXX:4:1111:3285:61573\t16\t1\t11980\t0\t124M\t*\t0\t0\tGGGTCTTGATGCTGCGGTCTTCATCTGCAGGTGTCTGACTTCCAGCAACTGCTGGCCTGTGCCAGGGTGCAAGCTGAGCACTGGAGTGGAGTTTTCCTGTGGAGAGGAGCCATGCCTAGAGTGG\t]]QVZ]]]]]L]]Z]]]]]]]]]]]]ZZ]]]]]]WLGW]QW=]]]\\]]]]]W]QZ]]]QQGG]]V]]ZVW]]+G]L]QWQW]L>LQV:VQ]]ZZ]]]:]]>\\]]]\\]V]]]]]=Q]:]]VV]QG\tXT:A:R\tNM:i:1\tX0:i:2\tX1:i:5\tXM:i:1\tXO:i:0\tXG:i:0\tMD:Z:14T109\tXP:i:1  \n",
    "    \n",
    "    M_ST-E00201:191:HH7YYALXX:4:2220:7872:9044\t0\t1\t13686\t0\t62M\t*\t0\t0\tGGCCATCCGTGAGATCTTCCCAGGGCAGCTCCCCTCTGTGGAATCCAATCTGTCTTCCATCC\t]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]\tXT:A:R\tNM:i:0\tX0:i:6\tX1:i:1\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:62\tXP:i:1  \n",
    "    \n",
    "    M_ST-E00201:191:HH7YYALXX:4:1222:4594:49074\t16\t1\t54317\t0\t41M\t*\t0\t0\tCTGCTTTGTAAATTATAACAGCTCAATTAAGAGAAACCGTA\t>]]]]]]]]]]]]]]]]]]]]]]]]]]]Z]]Z]]]]]V]]]\tXT:A:R\tNM:i:0\tX0:i:3\tX1:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:41\tXA:Z:15,+102476791,41M,0;19,-95927,41M,0;\tXP:i:1  \n",
    "    \n",
    "    M_ST-E00201:191:HH7YYALXX:4:2107:25986:14547\t0\t1\t66678\t20\t100M\t*\t0\t0\tACCTGTCAACAGTGGCTGGCTCTTCATGGTTGCTACAATGAGTGTGTAAGATTCTGAAGGACTCCTTTAATAAGCCTAAACTTAATGTTCAACTTAGAAT\t]Q]]]]Z]]Z]]]]]Z]]]]]Z]]]]]Z]]]]]]]Z]]]]]]]\\]]]]]V]]]]]Z]]V]]V]]Q]]]]]]]]]]]]]]]]]]]]]Z]]]]]]]]]]QV]\tXT:A:U\tNM:i:0\tX0:i:1\tX1:i:2\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:100\tXA:Z:15,-102465140,100M,1;19,+108266,100M,1;\tXP:i:1  \n",
    "    \n",
    "    * last 5 lines for jar001:  \n",
    "    `samtools view jar001-b1e1l1p1_CATGCTC_L004_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam | tail -n 5`  \n",
    "    M_ST-E00201:191:HH7YYALXX:4:2113:28239:31142\t16\ths37d5\t35412866\t37\t57M\t*\t0\t0\tCACTCCCCTCCTCCCAGTTTTACTCCACTCTTTTCCATTCCATTCCATTGCCCTCCA\t]]]]V]]V]]]]]]]]]]]]]]Z]Z]]]]]]]]]ZZ]]]]]]]]]]]]]]]]]]]]]\tXT:A:U\tNM:i:0\tX0:i:1\tX1:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:57\tXP:i:1  \n",
    "    \n",
    "    M_ST-E00201:191:HH7YYALXX:4:1215:25489:32707\t0\ths37d5\t35415637\t37\t128M\t*\t0\t0\tATTCCATTCAATCCCATTCCACTCTGGTTGATTCCATTCCATTCCATTCCTTCCCATTCCTATCCATTCCTTTCCATTCCATTCCATTCCATTCCATTCCATTCCATTCCCCTCAGGTTGATTCAATT\t]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]Z]]]]>]]]]Z]]]]]]]]]]>]]]Z]]]]]]]\tXT:A:U\tNM:i:0\tX0:i:1\tX1:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:128\tXP:i:1  \n",
    "    \n",
    "    M_ST-E00201:191:HH7YYALXX:4:2219:8532:42991\t16\ths37d5\t35424374\t37\t90M\t*\t0\t0\tGCCAGTGGATCTTAGCTTGCTGTGCTCCATGGGGGTGGGATCTGCTGAGCTAGACGACTTGGCTCCCTGGCTTCAGCCTCCTTTCCAGGG\t]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]\tXT:A:U\tNM:i:0\tX0:i:1\tX1:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:90\tXP:i:1  \n",
    "    \n",
    "    M_ST-E00201:191:HH7YYALXX:4:1118:16488:12894\t16\ths37d5\t35430325\t0\t45M\t*\t0\t0\tCCATTCCATTCCATTCCATTCCATTCCTTTCCATTCCGTTCCATT\t]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]\tXT:A:R\tNM:i:0\tX0:i:14\tX1:i:699\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:45XP:i:1  \n",
    "    \n",
    "    M_ST-E00201:191:HH7YYALXX:4:2109:23013:20331\t16\ths37d5\t35445116\t37\t117M\t*\t0\t0\tATCTTATGTACAGTATAAAATCTATATTTGATGTACTTTCATATTGTGTGTACACAATATAATATATATTGAATGTAAGTTCATATTTTATGTACAGTATATAATGTATAGTTTGGA\t]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]Z]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]\tXT:A:U\tNM:i:2\tX0:i:1\tX1:i:0\tXM:i:2\tXO:i:0\tXG:i:0\tMD:Z:74A41G0\tXP:i:1  \n",
    "    \n",
    "    * first 5 lines for jar002:  \n",
    "    `samtools view jar002-b1e1l1p1_ACGCAAC_L005_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam| head -n 5`  \n",
    "    M_ST-E00201:191:HH7YYALXX:5:2218:27478:34922\t16\t1\t23141\t0\t52M\t*\t0\t0\tCTTCCTAGGTGCCAGGCACTGTTCCATTCCTTTGCATGTTTTGATTAATTTA\t]Z]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]\tXT:A:R\tNM:i:0\tX0:i:4\tX1:i:1\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:52XP:i:1  \n",
    "    \n",
    "    M_ST-E00201:191:HH7YYALXX:5:1104:30614:57196\t0\t1\t79458\t0\t35M\t*\t0\t0\tGGAAAGAAACAAAGCGCCTGAAGGCTGCACCTGAA\t]Q]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]\tXT:A:R\tNM:i:0\tX0:i:4\tX1:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:35\tXA:Z:6,+69462,35M,0;19,+121055,35M,0;15,-102452843,35M,0;\tXP:i:1  \n",
    "    \n",
    "    M_ST-E00201:191:HH7YYALXX:5:1215:27752:67357\t0\t1\t110684\t0\t82M\t*\t0\t0\tAGGAACAGAAAACCAAATACAGCATACTCTCAGTTATAAGTGGGAGCTAAATGATGAGAACTCATGAAAACAAAGAATAAAA\t]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]Z]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]\tXT:A:R\tNM:i:1\tX0:i:6\tX1:i:9\tXM:i:1\tXO:i:0\tXG:i:0\tMD:Z:68C13\tXP:i:1  \n",
    "    \n",
    "    M_ST-E00201:191:HH7YYALXX:5:2206:10115:60817\t0\t1\t137197\t0\t78M\t*\t0\t0\tCTCTGTCCGCGTGGGAGGGGCCGGTGTGAGGCAAGGGCTCACACTGACCTCTCTCAGCGTGGGAGGAGCCAGTGTGAG\t]\\V]]ZZ]]]]]]]]]]]]]]]]]Z]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]\\]]]]]\tXT:A:R\tNM:i:0\tX0:i:5\tX1:i:2\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:78\tXP:i:1  \n",
    "    \n",
    "    M_ST-E00201:191:HH7YYALXX:5:2215:25367:21034\t0\t1\t245495\t0\t71M\t*\t0\t0\tTCTGCCTCCCAGGTTCAAGTGATTCTCCTGCCTCAGCCGCCCAAGTAGCTGGGATTACAGGTGTCTGCCAC\t]\\]]]]ZZZ]]]]]]]]]]]]Z]]Z]]]]]Z]]]]]]]]]]]]]]]]]Z]]]]]]]]]]]]]Z]]]]]]]]\tXT:A:R\tNM:i:0\tX0:i:6\tX1:i:9\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:71\tXP:i:1  \n",
    "    \n",
    "    * last 5 lines for jar002:  \n",
    "    `samtools view jar002-b1e1l1p1_ACGCAAC_L005_merged.170314_ST-E00201_0191_AHH7YYALXX.hs37d5.fa.cons.90perc.bam| tail -n 5`  \n",
    "    M_ST-E00201:191:HH7YYALXX:5:2204:23673:20172\t0\ths37d5\t35333000\t37\t48M\t*\t0\t0\tCTCCATTCCACTCCACTGCACTACTCCCCACTCCACTCCATTCCATTC\t]]]]]]]]]]]]]Z]]]]]]]]]]]]]]]]]]]]Z]]]]]]]]]]]]]\tXT:A:U\tNM:i:0\tX0:i:1\tX1:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:48\tXP:i:1  \n",
    "    \n",
    "    M_ST-E00201:191:HH7YYALXX:5:1211:2615:15268\t16\ths37d5\t35349167\t0\t45M\t*\t0\t0\tGCTTACAGGGGCTATTGTGACATATCTCTGCACTGATCACCCAGG\t]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]\tXT:A:R\tNM:i:1\tX0:i:2\tX1:i:46\tXM:i:1\tXO:i:0\tXG:i:0\tMD:Z:41T3\tXP:i:3  \n",
    "    \n",
    "    M_ST-E00201:191:HH7YYALXX:5:2119:5477:55350\t0\ths37d5\t35437046\t37\t49M\t*\t0\t0\tTCATTCAAGTCCTGTCCATTCCATTCCTCTCCATTCCACTCCACTCCAC\t]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]\tXT:A:U\tNM:i:2\tX0:i:1\tX1:i:0\tXM:i:2\tXO:i:0\tXG:i:0\tMD:Z:0C42G5\tXP:i:2  \n",
    "    \n",
    "    M_ST-E00201:191:HH7YYALXX:5:2216:4929:46947\t0\ths37d5\t35449519\t37\t127M\t*\t0\t0\tCACTGAGGCTGCAGTCCGGTGGCCCAAACTCAATTAACATTTCATTTCACCCTTCCATTCCATTGCGTTCCATTTCATTCCATTCCATTTCATTCCATTGCATTCCATTCCATTAAATTCCATTGCA\tG]\\W]W]]]]Z]]]Q]]]]QZ]]]]]]\\]W]]]Q]]]]]]]]]]]]VV]]Z]\\QWQV]]]]]]V]\\V]V]Z]]]]]]]]]]V]]]]Z]]ZVZ]Z]Z]]]]]]]]]QZ]VQ]Z]VQ]]]VVWV\\W=Q]\tXT:A:U\tNM:i:0\tX0:i:1\tX1:i:0\tXM:i:0\tXO:i:0\tXG:i:0\tMD:Z:127\tXP:i:1  \n",
    "    \n",
    "    M_ST-E00201:191:HH7YYALXX:5:1107:2879:6829\t0\ths37d5\t35461699\t37\t120M\t*\t0\t0\tATAGTGACGGTGATGAGGATTAGGATGATGGTGATGAGGATAGTGATGATGGTCATGGTGATGATGGTGATGAGGATGGTGATGATGGTGATAGTGATGGTGATGATGGTAATGATGATG\tQ\\V]GZVW]]G]L=W\\V]]=W]]]V0V]=]]VGV5]L]LV=VVL]Z>]V5]ZQ+:QQ]V]]LQG]]VL]]]]G]0G=]ZV]Q=]VLQ]]]]]]]Z]]V]]Z]]]]ZZ]VVV]G]Z]]]]]\tXT:A:U\tNM:i:1\tX0:i:1\tX1:i:0\tXM:i:1\tXO:i:0\tXG:i:0\tMD:Z:7T112\tXP:i:1   \n",
    "\n",
    "* the sequencing lists after `mpileup` extraction:  \n",
    "    * count the number of lines:  \n",
    "    `less jar001.mpileup |wc -l`  \n",
    "    or  \n",
    "    `less jar001.mpileup |awk 'END { print NR }'`  \n",
    "    53 588 203 - for 1st  \n",
    "\n",
    "    `less jar002.mpileup |wc -l`  \n",
    "    or  \n",
    "    `less jar002.mpileup |awk 'END { print NR }'`  \n",
    "    5 692 180 - for 2nd   \n",
    "    * print first few lines:  \n",
    "    `less -c jar001.mpileup |head -n 5`  \n",
    "    1\t10022\tC\t1\t^!,\tA  \n",
    "    1\t10023\tC\t1\t,\tB  \n",
    "    1\t10024\tC\t1\t,\tE  \n",
    "    1\t10025\tT\t1\t,\t]  \n",
    "    1\t10026\tA\t1\t,\t]  \n",
    "\n",
    "    `less -c jar001.mpileup |tail -n 5`  \n",
    "    hs37d5\t35445228\tt\t1\t,\tZ  \n",
    "    hs37d5\t35445229\tt\t1\t,\tR  \n",
    "    hs37d5\t35445230\tg\t1\t,\t9  \n",
    "    hs37d5\t35445231\tg\t1\t,\t1  \n",
    "    hs37d5\t35445232\tg\t0\t*\t*  \n",
    "\n",
    "    `less -c jar002.mpileup |head -n 5`  \n",
    "    1\t23141\tC\t1\t^!,\tE  \n",
    "    1\t23142\tT\t1\t,\tU  \n",
    "    1\t23143\tT\t1\t,\t]  \n",
    "    1\t23144\tC\t1\t,\t]  \n",
    "    1\t23145\tC\t1\t,\t]  \n",
    "\n",
    "    `less -c jar002.mpileup |tail -n 5`  \n",
    "    hs37d5\t35461814\tt\t1\t.\t]  \n",
    "    hs37d5\t35461815\tg\t1\t.\t]  \n",
    "    hs37d5\t35461816\ta\t1\t.\t]  \n",
    "    hs37d5\t35461817\tt\t1\t.\tT  \n",
    "    hs37d5\t35461818\tg\t1\t.$\tE  \n",
    "\n",
    "* make smaller files to use for making the script (5000 lines):  \n",
    "    `less jar001.mpileup |sed -n '1,5000p' > jar001.mpileup_5klines`  \n",
    "    `less jar002.mpileup |sed -n '1,5000p' > jar002.mpileup_5klines`  \n",
    "    * figure out the last of the loc & position of each file, then crop the list with genomic positions exactly after that loc & position:  \n",
    "        * last position in jar001.mpileup is:  \n",
    "        `less jar001.mpileup_5klines | tail -n 1`  \n",
    "        1\t416732\tA\t1\t.\t\\]  \n",
    "        * last position in jar002.mpileup is:  \n",
    "        1\t2839409\tA\t1\t.\t\\]  \n",
    "        * figure out the position in the genomic document that is one bigger than the biggest one in the sequences (bigger than the last position in jar002, 2 839 409):  \n",
    "        `less sima_informative_sites.tab |awk '{if ($1=1 && 2839400<$2 && $2 < 2839430) print NR \" - \" $0;}'`  \n",
    "        2433871 - 1 2839413 A G all  \n",
    "        * (for jar001 it is 416732:  \n",
    "        `less sima_informative_sites.tab |awk '{if ($1=1 && 416725<$2 && $2 < 416760) print NR \" - \" $0;}'`  \n",
    "        1951787 - 1 416757 C G all)  \n",
    "        * copy the positions including 2 839 409 from the genomic positions to a new file:  \n",
    "        `less sima_informative_sites.tab | sed -n '1,2433871p'>sima_crop.tab`  \n",
    "            * the table isn't sorted ('-n' numerical sort; '-k' sort by column 2):  \n",
    "            `less sima_informative_sites.tab |sort -n -k2 > sima_sort3`  \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### F) Write the python code:  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###  more usefull code:  \n",
    "* find out if I have a certain duplication in 'pos' column:  \n",
    "`less sima_informative_sites.tab_sort | awk '{if ($2 == \"179281355\") print $0;}'`  \n",
    "  \n",
    "* print out the lines that are duplicate in columns 1 and 2:  \n",
    "`less sima_informative_sites.tab_sort | awk 'FNR==NR{a[$1,$2]++;next} a[$1,$2]>1'`  \n",
    "\n",
    "* crop a large file to the first 5000 lines  \n",
    "`'less jar001.mpileup |sed -n '1,5000p' > jar001.mpileup_5klines'`  \n",
    "  \n",
    "* print the number of the line that matches the query:  \n",
    "`less duplicated_rows_tab| awk '{if ($2 == \"179281355\") print NR\" : \"$0;}'`"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
